src/main/java/org/reactome/addlinks/dataretrieval/ensembl/EnsemblBiomartRetriever.java

```diff
@@ -0,0 +1,232 @@
+                    // Since we need the ID, we require the List to have 4 elements.
+                    if (tabSplit.size() == 4) {
+                        microarrayIdentifierLines.add(microarrayResponseLine + "\n");
+                    }
+                }
+
+                // Write data to file.
+                if (!microarrayIdentifierLines.isEmpty()) {
+                    File speciesMicroarrayFile = createNewFile(this.destination + biomartSpeciesName + "_microarray");
+                    for (String probeIdLine : microarrayIdentifierLines) {
+                        Files.write(Paths.get(speciesMicroarrayFile.toString()), probeIdLine.getBytes(), StandardOpenOption.APPEND);
+                    }
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+    }
+
+    /**
+     * Attempts to create a connection to Biomart with the supplied query
+     * @param queryString String of Biomart query
+     * @return Set<String> Biomart response data.
+     * @throws Exception Can be caused by the query URL not being valid, or the data not existing for the species.
+     */
+    private Set<String> queryBiomart(String queryString, int retryCount) throws Exception {
+        // Create connection to Biomart URL for each species, retrieving a list of microarray probe types, if available.
+        URL biomartUrlWithSpecies = new URL(queryString);
+        HttpURLConnection biomartConnection = (HttpURLConnection) biomartUrlWithSpecies.openConnection();
+        if (biomartConnection.getResponseCode() == 200) {
+            Set<String> biomartResponseLines = new HashSet<>();
+            BufferedReader br = new BufferedReader(new InputStreamReader(biomartConnection.getInputStream()));
+            String line;
+            while ((line = br.readLine()) != null) {
+                // Biomart still responds with a 200, even if no data exists. For now, we handle it by
+                // checking the returned content for the string 'ERROR'. It will retry up to 5 times, with a 10 second delay.
+                if (line.contains("ERROR") ) {
+                    retryCount++;
+                    if (retryCount < 5) {
+                        Thread.sleep(10000);
+                        logger.warn("Biomart query failed. Trying again...");
+                        return queryBiomart(queryString, retryCount);
+                    }
+                    throw new Exception("Biomart query failed with message: " + line);
+                } else {
+                    biomartResponseLines.add(line);
+                }
+            }
+            return biomartResponseLines;
+
+        } else {
+            retryCount++;
+            if (retryCount < 5) {
+                Thread.sleep(10000);
+                logger.warn("Unable to connect to Biomart. Trying again...");
+                return queryBiomart(queryString, retryCount);
+            }
```

Comment:
This logic is identical to the retry block above.  You could extract this to a helper method like this:

```
private Set<String> retryQuery(String queryString, int retryCount, String retryMessage)
    throws InterruptedException, IOException, HttpException, BioMartQueryException {

    final long QUERY_SLEEP_DURATION = 10;

    Thread.sleep(Duration.ofSeconds(QUERY_SLEEP_DURATION).toMillis());
    logger.warn(retryMessage);
    return queryBiomart(queryString, retryCount);
}
```

Using the `Duration.of()` @SolomonShorser-OICR recommended, the named constant may not be needed here.

Reply:
Still a good idea to use a constant if you want that value to be uniformly used in multiple places. But then you could just create a private static member.

```Java
private static long QUERY_SLEEP_DURATION = Duration.OfSeconds(10).toMillis();
...
Thread.sleep(QUERY_SLEEP_DURATION);
```

Reply:
@SolomonShorser-OICR Sounds good to me.  A named constant if it's used more than one place in the class makes sense.  If it's contained only within one method, I think using the constant is more a judgement call.